src/search.rs

```diff
@@ -188,7 +186,7 @@ impl ImplReducer {
                 }
```

Comment:
### This patch's problem

Currently, this patch panics when parallelism is enabled. A single thread works as expected. The panic is caused by `src/db.rs:492:37`, e.g., `Missed sub-Impl (...) while computing cost for (...)`. I think the reason is the non-deterministic insertion and dropping in the `ImplReducer`.

### `ImplReducer` Problems

#### 1. Non-deterministic ordering with the same costs

When we have the same cost, which one will be chosen by `binary_search_by_key` is undefined, IIUC. If we assume we can always choose the first one, then the behavior would be that the first inserted same cost stays the longest since we insert after the found cost and pop the last one. I guess this weird ordering caused inconsistency with parallelism. I think we should define deterministic ordering that is robust even with parallelism. Or is this intended?

```rs
let reducer = ImplReducer::new(3); // []
reducer.insert(CostA(1)); // [CostA(1)]
reducer.insert(CostB(1)); // [CostA(1), CostB(1)]
reducer.insert(CostC(1)); // [CostA(1), CostB(1), CostC(1)]
reducer.insert(CostD(1)); // [CostA(1), CostD(1), CostB(1)] -poped> CostC(1)
reducer.insert(CostE(1)); // [CostA(1), CostE(1), CostD(1)] -poped> CostB(1)
reducer.insert(CostF(1)); // [CostA(1), CostF(1), CostE(1)] -poped> CostD(1)
```

#### 2. (Not related to the panic) Performance

Since we do `insert` over `self.results`, the time complexity is `O(K)`, where K is the top K. By using max-heap, we can reduce this to `O(log K)`. But the `finalize` method will take `O(K log K)`. Considering `insert` will be called over S specs and A actions and `finalize` will be called over S specs, I thought `O(S * A * K)` -> `O(S * A * log K)` and `O(S * 1)` -> `O(S * K log K)` changes might worth. Or since K is `u8`, which is at most `255`, do we consider `O(K)` of `insert` is negligible?

Reply:
**1.** At least in recent Rusts, [binary_search_by_key](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search_by_key) is deterministic but arbitrary, but you're right that different workers will update in different orders and it is important results not change when the number of parallel workers does! One way to solve this would be to sort by (Cost, action_index).

**2.** Optimizing for larger k is a good idea. We would really like to be able to support k up to 255 (at least), so the asymptotic behavior is important. I would just take care not introduce too much overhead in the common `k=1` case (e.g., from additional heap allocations).

However, I don't yet understand your analysis. Three questions:
- What Specs does `S` count? Can't that be left out of this analysis entirely, since there is one reduction per Spec?
- Isn't the current insertion time is `O(K log K)`, since there's an `O(log K)` search for the insertion point and then an `O(K)` shift of the remainder of the array?
- What work is `finalize` doing such that it takes `O(K log K)`? Shuffling the binary tree into a sorted Vec? If so, can we avoid this by returning binary trees rather than sorted Vecs?

Reply:
**1.** I didn't know it was deterministic, thank you! I will try `(Cost, action_idx)`.

> However, I don't quite see how this would cause the above panic.

Given a single thread worked well and only `top_k > 1` with parallelism fails, I guessed a needed action was accidentally dropped through `ImplReducer`. But I'm sure I need more investigation.

**2.** For the overhead, AFAIK, there's no `SmallVec`-like crate for `BinaryHeap`, we always do need one-time allocation (we do not need more allocation actions since we can do `::with_capacity(top_k)`). Allocation cost can be considered by the number and size of allocations, I think. Obviously, we have 0 -> 1 number of allocations compared to SmallVec, but I am not sure this would be significant. What do you think?

1. The reason why I included `S` in my analysis is for perspective from the entire program run. Because both `insert` and `finalize` have `S`, I think we can simply say `O(A * K)` and `O(1)`, respectively. In this case, it is like analyzing the complexity from top_down? I think the difference is where we start thinking about complexity, so I'll go with the perspective you feel most comfortable with!
2. For each insertion, the binary search is run once and shift also once. So, I think `O(log K + K)` = `O(K)`.
3. When we want to keep the return type for `finalize`, we want to return a sorted Vec. We can use [`into_sorted_vec`](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#method.into_sorted_vec), but which is (I think) equivalent to calling `pop()` over `K` elements, resulting in `O(K log K)`.

> If so, can we avoid this by returning binary trees rather than sorted Vecs?

I still do not fully understand `Database::put`, but if `NDArray::fill_region` assumes the 2nd argument is sorted, we cannot avoid `O(K log K)`. We can return `BinaryTree` itself, but the iterator [does not return elements in any specific order](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#method.iter). So, we need to call [`into_iter_sorted`](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#method.into_iter_sorted) to get a sorted result, which is basically the same as `into_sorted_vec`. But if `NDArray::fill_region` does not require a sorted Vec, we can simply use the iterator, leading to `O(1)` for `finalize`.

Reply:
> **2.** For the overhead, AFAIK, there's no `SmallVec`-like crate for `BinaryHeap`, we always do need one-time allocation (we do not need more allocation actions since we can do `::with_capacity(top_k)`). Allocation cost can be considered by the number and size of allocations, I think. Obviously, we have 0 -> 1 number of allocations compared to SmallVec, but I am not sure this would be significant. What do you think?

I'm not sure if it'll be significant either. If it is, we can deal with that after the implementation is correct.

So the runtimes per top_down call/reduction are:
| | Now | min-heap Proposal |
|--------|--------|--------|
| `insert` (run `A` times) | O(A * K) | O(A * log K) |
| `finalize` (run once) | O(1) | O(K * log K) | 
| Total | O(A * K) |  O((A + K) * log K) 

Looks like a win in most cases. I say: go for it. (We can do something cleverer later if this turns out to be a place we spend a lot of compute time.)

Reply:
> > **2.** For the overhead, AFAIK, there's no `SmallVec`-like crate for `BinaryHeap`, we always do need one-time allocation (we do not need more allocation actions since we can do `::with_capacity(top_k)`). Allocation cost can be considered by the number and size of allocations, I think. Obviously, we have 0 -> 1 number of allocations compared to SmallVec, but I am not sure this would be significant. What do you think?
> 
> I'm not sure if it'll be significant either. If it is, we can deal with that after the implementation is correct.
> 

I see.  It might be worth implementing `SmallVec`-like `BinaryTree` if it is.

> So the runtimes per top_down call/reduction are:
> 
> Now	min-heap Proposal
> `insert` (run `A` times)	O(A * K)	O(A * log K)
> `finalize` (run once)	O(1)	O(K * log K)
> Total	O(A * K)	O((A + K) * log K)

Just a nitpicking, but I think we should use max-heap, not min-heap, because we want to pop the largest cost when `reducer.len() >= top_k`.

> Looks like a win in most cases. I say: go for it. We can do something cleverer (maybe a selection algo.?) later if this turns out to be a place we spend a lot of compute time.

Thank you!